File: app/services/token_usage.py
"""Token usage estimation utilities (no external tokenizer)."""

from __future__ import annotations

from typing import Any, Dict, List, Tuple


def _count_tokens(text: str) -> int:
    if not text:
        return 0
    ascii_count = 0
    non_ascii = 0
    for ch in text:
        if ord(ch) <= 0x7F:
            ascii_count += 1
        else:
            non_ascii += 1
    return (ascii_count + 3) // 4 + non_ascii


def _split_think(text: str) -> Tuple[str, str]:
    if not text:
        return "", ""
    reasoning_parts: List[str] = []
    output = text
    start = 0
    while True:
        s = output.find("<think>", start)
        if s < 0:
            break
        e = output.find("</think>", s + 7)
        if e < 0:
            break
        reasoning_parts.append(output[s + 7 : e])
        output = output[:s] + output[e + 8 :]
        start = s
    return "\n".join(reasoning_parts), output


def estimate_prompt_tokens(messages: List[Dict[str, Any]]) -> Dict[str, int]:
    text_parts: List[str] = []
    image_tokens = 0
    for msg in messages or []:
        content = msg.get("content")
        if isinstance(content, list):
            for item in content:
                if isinstance(item, dict) and item.get("type") == "text":
                    t = str(item.get("text") or "")
                    if t.strip():
                        text_parts.append(t)
        else:
            t = str(content or "")
            if t.strip():
                text_parts.append(t)
    text_tokens = _count_tokens("\n".join(text_parts))
    return {
        "text_tokens": text_tokens,
        "image_tokens": image_tokens,
        "prompt_tokens": text_tokens + image_tokens,
    }


def build_chat_usage(messages: List[Dict[str, Any]], completion_text: str) -> Dict[str, Any]:
    prompt = estimate_prompt_tokens(messages)
    reasoning_text, output_text = _split_think(completion_text or "")
    completion_text_tokens = _count_tokens(output_text)
    reasoning_tokens = _count_tokens(reasoning_text)
    output_tokens = completion_text_tokens + reasoning_tokens
    input_tokens = prompt["prompt_tokens"]
    total_tokens = input_tokens + output_tokens
    return {
        "prompt_tokens": input_tokens,
        "completion_tokens": output_tokens,
        "total_tokens": total_tokens,
        "prompt_tokens_details": {
            "cached_tokens": 0,
            "text_tokens": prompt["text_tokens"],
            "audio_tokens": 0,
            "image_tokens": prompt["image_tokens"],
        },
        "completion_tokens_details": {
            "text_tokens": completion_text_tokens,
            "audio_tokens": 0,
            "reasoning_tokens": reasoning_tokens,
        },
        "_raw": {
            "reasoning_tokens": reasoning_tokens,
            "cached_tokens": 0,
            "input_tokens": input_tokens,
            "output_tokens": output_tokens,
        },
    }


def build_image_usage(prompt: str, success_count: int = 1) -> Dict[str, Any]:
    text_tokens = _count_tokens(str(prompt or ""))
    success = max(1, int(success_count or 1))
    total_tokens = text_tokens * success
    return {
        "total_tokens": total_tokens,
        "input_tokens": total_tokens,
        "output_tokens": 0,
        "input_tokens_details": {"text_tokens": text_tokens, "image_tokens": 0},
        "_raw": {
            "reasoning_tokens": 0,
            "cached_tokens": 0,
            "input_tokens": total_tokens,
            "output_tokens": 0,
        },
    }

File: src/utils/token_usage.ts
export interface TokenUsageCounts {
  total_tokens: number;
  input_tokens: number;
  output_tokens: number;
  reasoning_tokens: number;
  cached_tokens: number;
  input_tokens_details: { text_tokens: number; image_tokens: number };
  output_tokens_details: { text_tokens: number; reasoning_tokens: number };
}

export function estimateTokens(text: string): number {
  const raw = String(text || "");
  if (!raw) return 0;
  let ascii = 0;
  let nonAscii = 0;
  for (const ch of raw) {
    if (ch.charCodeAt(0) <= 0x7f) ascii += 1;
    else nonAscii += 1;
  }
  const asciiTokens = Math.ceil(ascii / 4);
  return asciiTokens + nonAscii;
}

export function splitThinkSegments(text: string): { reasoningText: string; outputText: string } {
  const raw = String(text || "");
  if (!raw) return { reasoningText: "", outputText: "" };
  const regex = /<think>([\s\S]*?)<\/think>/gi;
  const reasoningParts: string[] = [];
  let output = raw;
  let match: RegExpExecArray | null;
  while ((match = regex.exec(raw))) {
    if (match[1]) reasoningParts.push(match[1]);
  }
  output = output.replace(regex, "");
  return { reasoningText: reasoningParts.join("\n"), outputText: output };
}

export function estimateInputTokensFromMessages(messages: Array<{ content?: unknown }>): {
  textTokens: number;
  imageTokens: number;
  promptTokens: number;
} {
  const parts: string[] = [];
  let imageTokens = 0;
  for (const msg of messages || []) {
    const content = (msg as any)?.content;
    if (Array.isArray(content)) {
      for (const item of content) {
        if (item?.type === "text" && typeof item.text === "string") parts.push(item.text);
        if (item?.type === "image_url") imageTokens += 0;
      }
    } else if (typeof content === "string") {
      parts.push(content);
    }
  }
  const textTokens = estimateTokens(parts.join("\n"));
  return { textTokens, imageTokens, promptTokens: textTokens + imageTokens };
}

export function buildChatUsageFromTexts(args: {
  promptTextTokens: number;
  promptImageTokens: number;
  completionText: string;
}): TokenUsageCounts {
  const { reasoningText, outputText } = splitThinkSegments(args.completionText);
  const completionTextTokens = estimateTokens(outputText);
  const reasoningTokens = estimateTokens(reasoningText);
  const outputTokens = completionTextTokens + reasoningTokens;
  const inputTokens = args.promptTextTokens + args.promptImageTokens;
  const totalTokens = inputTokens + outputTokens;
  return {
    total_tokens: totalTokens,
    input_tokens: inputTokens,
    output_tokens: outputTokens,
    reasoning_tokens: reasoningTokens,
    cached_tokens: 0,
    input_tokens_details: { text_tokens: args.promptTextTokens, image_tokens: args.promptImageTokens },
    output_tokens_details: { text_tokens: completionTextTokens, reasoning_tokens: reasoningTokens },
  };
}

export function buildImageUsageFromPrompt(prompt: string): TokenUsageCounts {
  const inputTokens = estimateTokens(prompt || "");
  return {
    total_tokens: inputTokens,
    input_tokens: inputTokens,
    output_tokens: 0,
    reasoning_tokens: 0,
    cached_tokens: 0,
    input_tokens_details: { text_tokens: inputTokens, image_tokens: 0 },
    output_tokens_details: { text_tokens: 0, reasoning_tokens: 0 },
  };
}